src/main.rs

```diff
@@ -33,25 +38,66 @@ impl Instr {
 }
 
 struct Interpreter {
-    pc: u8,
-    acc: u8,
+    pc: usize,
+    acc: u16,
     ram: Vec<u16>,
 }
 
 impl Interpreter {
-    fn execute(&self) {
-        println!("{}", self.pc);
+    fn decode(&mut self, register: usize) -> Result<u16, LMCError> {
+        self.ram
+            .get(register)
+            .map_or_else(|| Err(LMCError::NotEnoughRAM), |value| Ok(*value))
+    }
+
+    fn decode_map(
+        &mut self,
+        register: usize,
+        f: impl FnOnce(&mut u16) -> Result<(), LMCError>,
+    ) -> Result<(), LMCError> {
+        self.ram
+            .get_mut(register)
+            .map_or_else(|| Err(LMCError::NotEnoughRAM), f)
+    }
+
+    fn execute(&mut self) -> Result<(), LMCError> {
+        let (operator, operand) = self
+            .decode(self.pc)
+            .map(|instr| (instr / 100, instr % 100))?;
 
+        self.pc += 1;
+
+        match (operator, operand) {
```

Comment:
We could just combine `(operator, operand)` into `cir` and rename it on line 64 too